### PR TITLE
fix: joins now replay correctly

### PR DESCRIPTION
**Pull request by zzlk**

### CODE REVIEW

File: hydroflow/src/compiled/pull/half_join_state/mod.rs [hunk 1]
```diff
@@ -5,17 +5,25 @@ mod multiset2;
...
 pub use reduce::HalfJoinStateReduce;
 pub use set::HalfSetJoinState;
+use smallvec::SmallVec;
 
 pub type SetJoinState<Key, V1, V2> = (HalfSetJoinState<Key, V1, V2>, HalfSetJoinState<Key, V2, V1>);
 
 pub trait HalfJoinState<Key, ValBuild, ValProbe> {
     fn build(&mut self, k: Key, v: &ValBuild) -> bool;
-    fn probe(&mut self, k: &Key, v: &ValProbe);
+    fn probe(&mut self, k: &Key, v: &ValProbe) -> Option<(Key, ValProbe, ValBuild)>;
```
Comment (MingweiSamuel) [inline hydroflow/src/compiled/pull/half_join_state/mod.rs:22]: did this change improve the benchmarks?

**Review by MingweiSamuel — APPROVED** [review]

File: hydroflow/src/compiled/pull/half_join_state/mod.rs (hunk 1 above)
Comment (zzlk) [reply hydroflow/src/compiled/pull/half_join_state/mod.rs:22]: I didn't run the benchmarks. I'll pull it out into a separate change so we can see if it helps or not.

**Comment by zzlk** [comment]
with the new implementation, performance looks like this now:

```
micro/ops/join          time:   [126.09 µs 126.49 µs 126.99 µs]
                        change: [+7.1771% +11.832% +16.740%] (p = 0.00 < 0.05)
                        Performance has regressed.

micro/ops/crossjoin     time:   [153.17 µs 154.05 µs 155.64 µs]
                        change: [+2.6760% +4.1625% +6.4488%] (p = 0.00 < 0.05)
                        Performance has regressed.

micro/ops/anti_join     time:   [12.307 µs 12.332 µs 12.362 µs]
                        change: [-7.0392% -5.2791% -3.8479%] (p = 0.00 < 0.05)
                        Performance has improved.

symmetric_hash_join/no_match
                        time:   [119.02 µs 119.30 µs 119.63 µs]
                        change: [-8.0666% -7.6772% -7.3305%] (p = 0.00 < 0.05)
                        Performance has improved.

symmetric_hash_join/match_keys_diff_values
                        time:   [257.80 µs 258.16 µs 258.62 µs]
                        change: [+41.449% +44.386% +47.019%] (p = 0.00 < 0.05)
                        Performance has regressed.

symmetric_hash_join/match_keys_same_values
                        time:   [258.21 µs 259.91 µs 262.30 µs]
                        change: [+47.619% +48.363% +49.140%] (p = 0.00 < 0.05)
                        Performance has regressed.

symmetric_hash_join/zipf_keys_low_contention_unique_values
                        time:   [174.77 µs 176.23 µs 177.87 µs]
                        change: [+10.626% +14.420% +19.165%] (p = 0.00 < 0.05)
                        Performance has regressed.

symmetric_hash_join/zipf_keys_high_contention_unique_values
                        time:   [3.5006 ms 3.5199 ms 3.5450 ms]
                        change: [-1.3268% -0.6624% +0.2329%] (p = 0.07 > 0.05)
                        No change in performance detected.
                        ```

**Review by MingweiSamuel — APPROVED** [review]

**Comment by zzlk** [comment]
fixes https://github.com/hydro-project/hydroflow/issues/871

**Pull request merged**